ouroboros-consensus/test-util/Test/Util/Orphans/Arbitrary.hs

```diff
@@ -111,6 +129,129 @@ instance Arbitrary Seed where
+      ShrunkConnectionPlan 1 (ConnectionPlan [4,5,4,0,1])
+
+  2 * n BFT consistently fails with at least the following
+
+      *** Failed! Exception: 'FailureException ExceptionInForkedThread (ImmDbParseFailure (Right (SlotNo {unSlotNo = 10})) (DeserialiseFailure 0 "expected list len"))' (after 6 tests and 7 shrinks):
+      ShrunkConnectionPlan 7 (ConnectionPlan [2,2,4,9,1])
+
+      *** Failed! Exception: 'FailureException ExceptionInForkedThread (ImmDbTrailingData (Right (SlotNo {unSlotNo = 11})) "\131\NUL\130\134\NUL\129\130")' (after 1 test and 3 shrinks):
+      ShrunkConnectionPlan 3 (ConnectionPlan [8,2,2,6,1])
+
+      *** Failed! Falsifiable (after 6 tests and 11 shrinks):
+      ShrunkConnectionPlan 11 (ConnectionPlan [8,9,1])
+      ...
+      empty intersection of non-empty chains (one reaches slot 18 and contains 10 blocks): Genesis :> (genesis->4ad58028,1,1,[([(b378192f,0)],[(b,769),(a,231)])]) :> (4ad58028->3349ef5e,4,1,[([(a8667d6a,1)],[(b,117),(a,114)])]) :> (3349ef5e->686f1c7c,7,1,[]) :> (686f1c7c->0d1dedd3,9,0,[([(5ec9a7bb,0),(a8667d6a,0),(b378192f,1)],[(a,1362),(b,524)])]) :> (0d1dedd3->7ca06c32,10,1,[]) :> (7ca06c32->f0e64a71,12,0,[]) :> (f0e64a71->2d834e24,13,1,[([(5ec9a7bb,1),(add6ae49,0)],[(c,751),(a,725)])]) :> (2d834e24->c5186ebc,15,0,[([(add6ae49,1)],[(c,428),(b,96)])]) :> (c5186ebc->3c9a6932,16,1,[]) :> (3c9a6932->6612c365,18,0,[])
+
+-}
+
+  where
+    -- TODO how to not conflate number of blocks and number of slots?
+    n = fromEnum (maxRollbacks defaultSecurityParam) :: Int
+
+-- | Inclusive minimum bound of 'boundsConnectionPlanFinalDuration'.
+minBoundsConnectionPlanFinalDuration :: Int
+minBoundsConnectionPlanFinalDuration = 1
+
+-- | Inclusive bounds for the last duration in @'arbitrary' :: 'ConnectionPlan'@.
+--
+-- Smaller than 'boundsNumSlots' to bound cost of @shrink @'ShrunkConnectionPlan'@.
+--
+-- Must be at least 1 so that the last node to join has an opportuniy to catch up.
+-- (TODO NSF someone else please review that claim.)
+boundsConnectionPlanFinalDuration :: (Int, Int)
+boundsConnectionPlanFinalDuration = (minBoundsConnectionPlanFinalDuration, 10)
+
+-- | Does not shrink. See 'ShrunkConnectionPlan'.
+instance Arbitrary ConnectionPlan where
+  arbitrary = do
+      numDurs <- choose boundsNumCoreNodes
+      durs <- vectorOf (numDurs - 1) $ choose boundsConnectionPlanDelay
+      finalDur <- choose boundsConnectionPlanFinalDuration
+
+      pure $ ConnectionPlan (durs ++ [finalDur])
+  shrink = const []
+
+instance Arbitrary ShrunkConnectionPlan where
+  arbitrary = ShrunkConnectionPlan 0 <$> arbitrary
+
+  -- This shrink is thorough, but expensive. TODO be less thorough?
+  --
+  -- We shrink in stages, so that we can prune based on the following heuristic:
+  -- reducing some durations in a passing plan cannot cause it to fail.
+  shrink (ShrunkConnectionPlan shrinkStage connPlan) =
```

Comment:
Ignoring on your request.

Reply:
OK, I like where it's at as of the push on Tuesday; please let me know what you think. (Frankly, feels like overkill at the moment, for various reasons. It doesn't feel obvious how to shrink while preserving the "relevant" parts of the plan.)

Reply:
It does look like overkill :slightly_smiling_face:. `subcompositions` seems a bit too complex :exploding_head: